sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java

```diff
@@ -302,6 +309,82 @@ public static SqlOrderBy convertClusterByToOrderBy(SqlNode query, SqlNodeList cl
     );
   }
 
+  /**
+   * Return resolved clustered by columns specified to output columns.
+   * For example,
+   * EXPLAIN PLAN FOR
+   * INSERT INTO w000
+   * SELECT
+   *  TIME_PARSE("timestamp") AS __time,
+   *  page AS page_alias,
+   *  FLOOR("cost"),
+   *  country,
+   *  citName
+   * FROM ...
+   * PARTITIONED BY DAY
+   * CLUSTERED BY 1, 2, 3, cityName
+   *
+   * The above SQL will return the following clusteredBy columns: ["__time", "page_alias", "FLOOR(\"cost\")", cityName"]
+   * Any ordinal and expression specified in the CLUSTERED BY clause will resolve to the final output column name.
+   * @param clusteredByNodes  List of {@link SqlNode}s representing columns to be clustered by.
+   * @param sourceNode The select or order by source node.
+   */
+  @Nullable
+  public static List<String> resolveClusteredByColumnsToOutputColumns(SqlNodeList clusteredByNodes, SqlNode sourceNode)
```

Comment:
General question about logic here; we seem to constrain the types of objects, and their child members a lot here, wondering how we know the types for sure? If they are different types than expected, are we expected to fail earlier with better error messaging?

Reply:
Right, the function is only accounting for types that are allowed in a `CLUSTERED BY` clause as the syntax tree can be recursive and exhaustive - partly adopted from the [grammar](https://github.com/apache/druid/blob/master/sql/src/main/codegen/includes/insert.ftl#L37C19-L37C31). The logic here is taking a defensive approach, so if we expect any unexpected type along the way, we'd throw an exception. I've added some inline comments in the different branches in this code.

Also, please see the unit tests in `DruidSqlParserUtilsTest`, `CalciteInsertDmlTest.java` and `CalciteReplaceDmlTest.java`.